File: api/app/javascript/controllers/republish_controller.js
import { Controller } from "@hotwired/stimulus";
import { i18nTable, t } from "../lib/i18n";
import { toast } from "../lib/toast";

/**
 * The Republish dialog: one delay in minutes, where zero is "now".
 *
 * The label of the submit button says what the current value does. That label is the only thing
 * that tells the owner "now" from "later", thus it must follow each edit of the field.
 */
export default class extends Controller {
  static targets = ["form", "minutes", "label"];

  connect() {
    // ⚠️ The words come from the locale file, through the `data-admin-i18n` attribute. Read the
    // table here: it is synchronous, and `relabel` runs at each keystroke.
    this.words = i18nTable(this.element);
    // ⚠️ The attribute, and not the property: it is the value that the server rendered, thus a
    // submit can put it back after the owner picked another delay.
    this.defaultMinutes = this.minutesTarget.getAttribute("value");
    // ⚠️ It waits for the definition. Before the upgrade, `value` of the field is undefined, and the
    // label would then stay at the plain word for a value that is correct.
    customElements.whenDefined("wa-number-input").then(() => this.relabel());
  }

  /**
   * Writes what the submit button does into its label.
   *
   * A value outside the limits gets the plain word, because the delay is then unknown. The field is
   * `required` and it has a min and a max, thus the browser stops that submit.
   */
  relabel() {
    const minutes = this.minutes;

    if (minutes === null) {
      this.labelTarget.textContent = t(this.words, "plain");
    } else if (minutes === 0) {
      this.labelTarget.textContent = t(this.words, "now");
    } else {
      this.labelTarget.textContent = t(this.words, "scheduled", { count: minutes });
    }
  }

  /**
   * Submits the form from the field, because the submit button is in a slot outside that form.
   *
   * ⚠️ It stops the default action first. Thus one Enter gives one submit, and it does that whether
   * or not the browser finds the button through its `form` attribute.
   * @param {KeyboardEvent} event
   */
  submitOnEnter(event) {
    if (event.key !== "Enter") return;

    event.preventDefault();
    this.formTarget.requestSubmit();
  }

  /**
   * Starts the republish with fetch, closes the dialog, and puts the answer in a toast.
   *
   * ⚠️ **It stops the native submit.** A submit that goes through navigates the admin, and the page
   * below the dialog renders again. The owner opens this dialog from each admin page, thus that
   * navigation would take away a draft of the Social media page that nothing has saved.
   *
   * ⚠️ It sends the CSRF token in a header. The admin does not skip the forgery protection.
   * @param {SubmitEvent} event
   */
  async submit(event) {
    event.preventDefault();
    this.close();

    try {
      const response = await fetch(this.formTarget.action, {
        method: "POST",
        headers: { Accept: "application/json", ...this.csrfHeader },
        body: new FormData(this.formTarget),
      });
      const answer = await response.json().catch(() => ({}));
      toast(answer.message || t(this.words, "unreachable"), response.ok ? "success" : "danger");
    } catch {
      toast(t(this.words, "unreachable"), "danger");
    }
  }

  /**
   * Closes the dialog and puts the delay back to the value of the server.
   *
   * ⚠️ The dialog stays in the DOM after a submit, thus a delay that the owner picked one time
   * would still be there the next time that they open it.
   */
  close() {
    this.element.open = false;
    this.minutesTarget.value = this.defaultMinutes;
    this.relabel();
  }

  /**
   * The CSRF token of the page, as a header.
   *
   * ⚠️ `csrf_meta_tags` renders nothing where the forgery protection is off, which is the test
   * environment. Thus this gives an empty object there and the header is absent.
   * @returns {object}
   */
  get csrfHeader() {
    const token = document.querySelector("meta[name='csrf-token']")?.content;

    return token ? { "X-CSRF-Token": token } : {};
  }

  /**
   * The delay of the field, or null when it is empty or outside the limits of the markup.
   * @returns {number|null}
   */
  get minutes() {
    const value = this.minutesTarget.value;
    if (value === "") return null;

    const minutes = Number(value);
    const min = Number(this.minutesTarget.getAttribute("min"));
    const max = Number(this.minutesTarget.getAttribute("max"));
    if (!Number.isInteger(minutes) || minutes < min || minutes > max) return null;

    return minutes;
  }
}

File: api/app/controllers/admin/republish_controller.rb
module Admin
  # Builds and deploys the static web site again, from the Republish dialog of the nav. That dialog
  # is in the admin layout, thus each admin page can start a build.
  #
  # There is no page here, and there is no GET. The nav item opens the dialog, and the dialog posts
  # to this one action.
  class RepublishController < BaseController
    # The limits of the delay, in minutes. The dialog writes them into the number input, and this
    # action checks them again, because a client can send any value.
    # ⚠️ A delay of zero is "now", and it is the default. The dialog has no separate control for it:
    # the label of the submit button is what says which one the value gives.
    MIN_MINUTES = 0
    MAX_MINUTES = 60
    DEFAULT_MINUTES = 0

    # POST /republish
    def create
      minutes = delay_minutes
      return refuse(t("admin.republish.toast.out_of_range", min: MIN_MINUTES, max: MAX_MINUTES)) if minutes.nil?
      return build_now if minutes.zero?

      replaced = SiteBuildJob.schedule_in(minutes.minutes, SiteBuildJob::ADMIN_EVENT_TYPE)
      key = replaced ? "rescheduled" : "scheduled"
      accept(t("admin.republish.toast.#{key}", count: minutes))
    end

    private

    # Adds an immediate build to the queue, or says that one already started.
    #
    # The lock is the same one that POST /api/build takes. Thus a double click, and a click after a
    # curl, cannot start two Actions runs.
    #
    # ⚠️ It takes the lock before it cancels the scheduled build. In the other order, a click inside
    # the window of the lock would cancel that build and start nothing.
    def build_now
      unless SiteBuildJob.claim_trigger_lock
        seconds = SiteBuildJob::TRIGGER_LOCK_TTL.to_i
        return refuse(t("admin.republish.toast.already_started", count: seconds))
      end

      cancelled = SiteBuildJob.cancel_scheduled
      SiteBuildJob.perform_async(SiteBuildJob::ADMIN_EVENT_TYPE)
      accept(t("admin.republish.toast.#{cancelled ? "now_cancelled" : "now"}"))
    end

    # Reads the delay of the form.
    #
    # ⚠️ It matches the digits before it converts. `to_i` gives 0 for text, and 0 is "now", thus a
    # value with a mistake would start a build.
    # @return [Integer, nil] The number of minutes, or nil when the value is outside the limits.
    def delay_minutes
      value = params[:minutes].to_s.strip
      return nil unless value.match?(/\A\d+\z/)

      minutes = value.to_i
      minutes.between?(MIN_MINUTES, MAX_MINUTES) ? minutes : nil
    end

    # @param message [String] The words of the toast.
    def accept(message)
      render json: { message: message }
    end

    # @param message [String] The words of the toast.
    def refuse(message)
      render json: { message: message }, status: :unprocessable_content
    end
  end
end
